In [56]:
from pathlib import Path
import csv
import numpy as np
import pandas as pd
from mmaction.core import confusion_matrix

In [8]:
p_pkl_dir = Path('/data/hyogun/repos/haawron_mmaction2/data/features')
weight = 'k400'
task = 'sim2real'
probed_on = 'k400'
p_subtask_dir = p_pkl_dir / weight / task / probed_on
p_pkl_train = p_subtask_dir / 'train.pkl'
p_pkl_test = p_subtask_dir / 'test.pkl'
x_train = np.array(np.load(p_pkl_train, allow_pickle=True))
x_test = np.array(np.load(p_pkl_test, allow_pickle=True))
x_train.shape, x_test.shape

((23085, 768), (961, 768))

In [36]:
p_ann_dir = Path(f"/data/hyogun/repos/haawron_mmaction2/data/_filelists/{probed_on}{'/processed' if probed_on=='k400' else ''}")
ann_train = p_ann_dir / f'filelist_{probed_on}_train_open.txt'
ann_test = p_ann_dir / f'filelist_{probed_on}_test_open.txt'
with ann_train.open() as f1, ann_test.open() as f2:
    ann_train = np.array(list(csv.reader(f1, delimiter=' ')))
    ann_test = np.array(list(csv.reader(f2, delimiter=' ')))
print(ann_train.shape); print(ann_train[:5])

(23085, 3)
[['tap_dancing/--6q_33gNew_000132_000142' '300' '5']
 ['climbing_a_rope/--EaS9P7ZdQ_000013_000023' '300' '12']
 ['balloon_blowing/--Ntf6n-j9Q_000017_000027' '300' '12']
 ['javelin_throw/--_S9IDQPLg_000135_000145' '300' '2']
 ['balloon_blowing/--gx7yb1-x0_000298_000308' '300' '12']]


In [23]:
sim = x_train @ x_test.T  # [N_train, N_test]
idx_best = sim.argmax(axis=0)  # [N_test], for each test data find the best-matching train point
idx_best.shape, idx_best[:5]

((961,), array([21480,  7553, 11331,  3376, 17984]))

In [45]:
pred = np.take_along_axis(ann_train[:,-1], idx_best, axis=0).astype(int)
gt = ann_test[:,-1].astype(int)
np.vstack((pred[:5], gt[:5]))

array([[12, 11,  9, 12,  6],
       [ 8, 11,  9, 12,  6]])

In [48]:
conf = confusion_matrix(pred, gt)
conf

array([[ 58,   0,   0,   1,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [  0,  38,   2,   1,   0,   0,   1,   0,   0,   0,   0,   1,   0],
       [  1,   1,  60,   8,   0,   2,   0,   0,   0,   2,   0,   0,   2],
       [  0,   0,   3,  57,   1,   0,   1,   0,   1,   1,   0,   0,   0],
       [  0,   0,   0,   0,  13,   1,   0,   0,   0,   0,   0,   0,   1],
       [  0,   0,   0,   0,   1,  95,   0,   0,   1,   0,   0,   0,   1],
       [  0,   0,   0,   0,   0,   1,  90,   0,   1,   1,   0,   0,   4],
       [  0,   0,   0,   0,   0,   0,   0,  19,   0,   0,   0,   0,   0],
       [  0,   1,   1,   0,   1,   1,   0,   0,  51,   1,   2,   1,  12],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,  15,   0,   0,   0],
       [  1,   0,   1,   1,   0,   3,   1,   0,   0,   2,  31,   0,   3],
       [  0,   0,   0,   0,   0,   0,   0,   1,   1,   0,   0,  48,   0],
       [  0,   0,   1,   4,   2,   5,   1,   0,   3,   1,   4,   0, 290]])

In [55]:
recalls = conf.diagonal() / conf.sum(axis=1)
os_star = recalls[:-1].mean()
unk = recalls[-1]
H = 2 / (1/os_star + 1/unk)
print(
    f'''H:   {100*H:.1f}
OS*: {100*os_star:.1f}
UNK: {100*unk:.1f}'''
)

H:   91.2
OS*: 89.3
UNK: 93.2


# Altogether

In [4]:
from pathlib import Path
import csv
import numpy as np
import pandas as pd
from mmaction.core import confusion_matrix
import inspect

weights = ['tsm', 'i3d', 'k400-SVT', 'k400', 'h100m', 'ssv2', 'in1k']
tasks = {
    'hello': ['ucf', 'hmdb'],
    'ek100': ['P02', 'P04', 'P22'],
    'simnreal': ['k400', 'babel']
}
real_dataset_name_mapping = {
    'ucf': ['ucf101', 'ucf'],
    'hmdb': ['hmdb51', 'hmdb'],
    'P02': ['ek100', 'P02'],
    'P04': ['ek100', 'P04'],
    'P22': ['ek100', 'P22'],
    'k400': ['k400', 'k400'],
    'babel': ['babel', 'babel'],
}

rows = []
p_pkl_dir = Path('/data/hyogun/repos/haawron_mmaction2/data/features')
for weight in weights:
    for task in tasks:
        for probed_on in tasks[task]:
            p_subtask_dir = p_pkl_dir / weight / task / probed_on
            print(f'{weight} --> {probed_on}')
            p_pkl_train = p_subtask_dir / 'train.pkl'
            p_pkl_test = p_subtask_dir / 'test_merged.pkl'
            if not p_pkl_train.is_file() or not p_pkl_test.is_file():
                rows.append(['0', '0', '0'])
                print(p_subtask_dir)
                print('\tpassed\n')
                continue
            x_train = np.array(np.load(p_pkl_train, allow_pickle=True))
            x_test = np.array(np.load(p_pkl_test, allow_pickle=True))

            dataset_dirname, dataset_filelist_name = real_dataset_name_mapping[probed_on]
            p_ann_dir = Path(f"/data/hyogun/repos/haawron_mmaction2/data/_filelists/{dataset_dirname}{'/processed' if probed_on=='k400' else ''}")
            p_ann_train = p_ann_dir / f'filelist_{dataset_filelist_name}_train_open_all.txt'
            p_ann_test = p_ann_dir / f'filelist_{dataset_filelist_name}_test_merged_open_all.txt'
            with p_ann_train.open() as f1, p_ann_test.open() as f2:
                ann_train = np.array(list(csv.reader(f1, delimiter=' ')))
                ann_test = np.array(list(csv.reader(f2, delimiter=' ')))

            sim = x_train @ x_test.T  # [N_train, N_test]
            idxx_best = sim.argsort(axis=0)[::-1]  # [N_train, N_test]
            pred = np.take(ann_train[:,-1].astype(int), idxx_best[:5])  # [5, N_test]
            gt = ann_test[:,-1].astype(int)  # [N_test]
            conf = confusion_matrix(pred[0], gt)

            top1_acc = (pred[0] == gt).sum() / gt.shape[0]
            top5_acc = ((pred == gt[None,:]).sum(axis=0) > 0).sum() / gt.shape[0]
            den = conf.sum(axis=1)
            recalls = conf.diagonal()[den>0] / den[den>0]  # ek100 has 0-pop classes
            mca = recalls.mean()
            rows.append([f'{100*mca:.1f}', f'{100*top1_acc:.1f}', f'{100*top5_acc:.1f}'])

            with np.printoptions(linewidth=10000, threshold=np.inf):
                print(conf)
            print(
                inspect.cleandoc(
                    f'''
                    mca:  {100*mca:.1f}
                    top1: {100*top1_acc:.1f}
                    top5: {100*top5_acc:.1f}
                    '''
                ), '\n'
            )

            # idx_best = sim.argmax(axis=0)  # [N_test], for each test data find the best-matching train point

            # pred = np.take_along_axis(ann_train[:,-1], idx_best, axis=0).astype(int)
            # gt = ann_test[:,-1].astype(int)

            # conf = confusion_matrix(pred, gt)

            # top1 = conf.diagonal().sum() / conf.sum()
            # top5 = 
            # recalls = conf.diagonal() / conf.sum(axis=1)
            # rows.append([f'{100*H:.1f}', f'{100*os_star:.1f}', f'{100*unk:.1f}'])
print(rows)

tsm --> ucf
/data/hyogun/repos/haawron_mmaction2/data/features/tsm/hello/ucf
	passed

tsm --> hmdb
/data/hyogun/repos/haawron_mmaction2/data/features/tsm/hello/hmdb
	passed

tsm --> P02
/data/hyogun/repos/haawron_mmaction2/data/features/tsm/ek100/P02
	passed

tsm --> P04
/data/hyogun/repos/haawron_mmaction2/data/features/tsm/ek100/P04
	passed

tsm --> P22
/data/hyogun/repos/haawron_mmaction2/data/features/tsm/ek100/P22
	passed

tsm --> k400
/data/hyogun/repos/haawron_mmaction2/data/features/tsm/simnreal/k400
	passed

tsm --> babel
[[ 82  14   4   0   0   0   0   0   1   1   0   0   0   0   0   1   0   1   7   0]
 [ 23 187   7   3   0   0   0   1   0   0   0   0   0   0   1   0   0   7   1   0]
 [ 21   3   8   3   0   0   0   1   1  17   0   0   0   0   0   4   0   1   0   0]
 [ 18  29   4  10   0   0   0   2   0   1   0   0   2   0   0   0   2   7   2   0]
 [  9  17   6   3   1   0   0   0   1   4   0   0   0   0   0   1   0   0   2   0]
 [  1   0   0  13   0   1   0   1   0   3   0   

In [4]:
all_subtasks = np.array(sum(map(list, tasks.values()), []) + ['Average'])
colnames = pd.MultiIndex.from_product([all_subtasks, ['MCA', 'Top1', 'Top5']])
rownames = np.array(weights)
print(colnames.shape, rownames.shape)
average = np.array(rows).reshape(5, 7, -1).astype(float).mean(axis=1).round(1)
result = np.hstack((np.array(rows).reshape(5, -1), average))
df = pd.DataFrame(result, columns=colnames, index=rownames)
pd.set_option('display.colheader_justify', 'center')
with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.expand_frame_repr', None):
    print(df)
print()
print(df.to_latex(multicolumn_format='c'))

(24,) (5,)
          ucf                  hmdb              P02               P04               P22               k400             babel             Average            
          MCA    Top1   Top5   MCA   Top1  Top5  MCA   Top1  Top5  MCA   Top1  Top5  MCA   Top1  Top5  MCA   Top1  Top5  MCA   Top1  Top5   MCA    Top1  Top5
k400-SVT    4.1    4.0   16.0   4.5   4.0  24.5   1.5   6.5  22.2   1.3  19.6  31.5   2.5  20.2  64.8   3.7   9.3  22.5   5.6   4.4  33.0    3.3    9.7  30.6
k400      100.0  100.0  100.0  89.3  88.4  96.1  27.0  33.0  72.6  23.7  29.3  63.3  30.6  35.8  74.4  86.8  87.4  94.5  55.2  65.6  85.3   58.9   62.8  83.7
h100m       5.3    3.7   22.8   5.5  16.0  33.0   1.5   1.5  13.9   1.7   1.7  28.4   2.6   2.2  41.9   4.0   9.4  16.0   5.0  19.5  57.7    3.7    7.7  30.5
ssv2        6.1    4.9   15.5   5.8   4.9  20.7   1.5  20.7  50.0   1.2  16.9  39.6   2.3   5.0  66.6   3.2   2.4  16.8   5.0   1.6  48.6    3.6    8.1  36.8
in1k        4.7    4.0   12.6   3.5   3.3

{   } & MCA  & Top1 &  Top5 && MCA  & Top1 &  Top5 && MCA  &  Top1 &  Top5 && MCA  &  Top1 &  Top5 && MCA  &  Top1 &  Top5 &&  MCA  &  Top1 &  Top5 &&  MCA  &  Top1 &  Top5 &&   MCA   &  Top1 &  Top5 \\
\midrule
k400  &  5.8 &  7.0 &  24.7 &&  4.2 &  5.9 &  26.2 &&  1.1 &   9.6 &  35.5 &&  1.4 &   7.4 &  31.0 &&  2.0 &  14.7 &  47.3 &&  86.8 &  87.4 &  94.5 &&  55.2 &  65.6 &  85.3 &&   22.4  &  28.2 &  49.2 \\
h100m &  4.2 &  5.9 &  31.8 &&  4.7 &  4.4 &  19.4 &&  1.5 &  20.7 &  42.3 &&  1.3 &  16.0 &  41.9 &&  2.3 &  22.5 &  49.0 &&   4.0 &   9.4 &  16.0 &&   5.0 &  19.5 &  57.7 &&    3.3  &  14.1 &  36.9 \\
ssv2  &  4.5 &  5.3 &  26.4 &&  4.5 &  4.6 &  30.8 &&  1.5 &  20.8 &  45.4 &&  1.3 &   3.6 &   8.4 &&  2.3 &  23.1 &  53.6 &&   3.2 &   2.4 &  16.8 &&   5.0 &   1.6 &  48.6 &&    3.2  &   8.8 &  32.9 \\
in1k  &  4.3 &  4.5 &  22.2 &&  4.2 &  5.2 &  28.9 &&  1.4 &  18.4 &  33.8 &&  1.6 &  16.8 &  25.7 &&  2.2 &   4.7 &  27.5 &&   4.0 &   3.8 &  18.2 &&   5.0 &  19.5 &  19.5 &&    3.2  &  10.4 &  25.1 \\